trieste/ask_tell_optimization.py

```diff
@@ -437,14 +437,17 @@ def acquisition_state(self) -> StateType | None:
     @classmethod
     def dataset_len(cls, datasets: Mapping[Tag, Dataset]) -> int:
         """Helper method for inferring the global dataset size."""
-        dataset_lens = {
-            len(dataset.query_points)
+        dataset_lens = [
+            tf.shape(dataset.query_points)[0]
             for tag, dataset in datasets.items()
             if not LocalizedTag.from_tag(tag).is_local
-        }
-        if len(dataset_lens) != 1:
+        ]
+        if dataset_lens and tf.reduce_all(
+            [tf.equal(tensor, dataset_lens[0]) for tensor in dataset_lens]
+        ):
+            return int(next(iter(dataset_lens)))
```

Comment:
Suggest using `tf.unique`.
```suggestion
        unique_lens, unique_idxs = tf.unique(dataset_lens)
        if len(unique_idxs) == 1:
            return unique_lens[0]
```